tests/Unit/HtmlProcessor/CssToAttributeConverterTest.php

```diff
@@ -171,8 +171,8 @@ public function notMatchingCssToHtmlMappingDataProvider(): array
             'span.float' => ['<span style="float: right;">Hi</span>'],
             'float: none' => ['<table style="float: none;"></table>'],
             'p.border-spacing' => ['<p style="border-spacing: 5px;">Hi</p>'],
-            'height: auto' => ['<img src="logo.png" alt="" style="height: auto;">'],
-            'width: auto' => ['<img src="logo.png" alt="" style="width: auto;">'],
+            'height: auto' => ['<img src="logo.png" alt style="height: auto;">'],
+            'width: auto' => ['<img src="logo.png" alt style="width: auto;">'],
```

Comment:
Is this valid HTML?

Reply:
Depends on the DOCTYPE. It's not valid XHTML, it's valid HTML:
> Attributes are placed inside the start tag, and consist of a name and a value, separated by an "=" character. The attribute value can remain unquoted if it doesn't contain ASCII whitespace or any of " ' ` = < or >. Otherwise, it has to be quoted using either single or double quotes. The value, along with the "=" character, can be omitted altogether if the value is the empty string.

```html
<!-- empty attributes -->
<input name=address disabled>
<input name=address disabled="">

<!-- attributes with a value -->
<input name=address maxlength=200>
<input name=address maxlength='200'>
<input name=address maxlength="200">
```

https://html.spec.whatwg.org/multipage/introduction.html#intro-early-example

Reply:
I see no reason to change this, the RNIB recommendation is `alt=""`.  We can add a separate test for `alt`.  We should not change the tests to allow the code to pass; we should fix the code to pass the tests - unless the tests are demonstrably wrong ;)

Reply:
I’ve looked into this a bit more.

[Earlier versions of the HTML5 spec](https://dev.w3.org/html5/html-author/#empty-attr) permitted the empty attribute syntax only for Boolean attributes.  [Later versions](https://stackoverflow.com/questions/30748847/is-an-empty-class-attribute-valid-html#comment-49561145) still don’t explicitly state that it is allowed for non-Boolean attributes, but they no longer say it is not (the specific clause, “This syntax is permitted only for boolean attributes,” has been removed).

Given that email clients are some way behind in their support for the specs (if they weren’t there’d be no need for Emogrifier), and many are probably still parsing HTML as HTML 4.01 or earlier, I would be reticent about outputting non-Boolean attributes with the empty attribute syntax, as it may cause issues on some clients.

`Masterminds/html5-php` actually has [rules](https://github.com/Masterminds/html5-php/blob/2.7.0/src/HTML5/Serializer/OutputRules.php#L70) defining non-Boolean attributes to be output with `=""` but these do not kick in when the `disable_html_ns` option is set.  But if that option is not set, an `xmlns` attribute is set on the root `html` element and our XPath expressions fail to match elements by type.

It would be possible to extend `OutputRules` to achieve the desired behaviour (which could also help in other cases, e.g. preventing `"\r"` in the output on Windows); to do so would also require extending `HTML5` to replace the `save` method to instantiate an extended `OutputRules`.

I note that the test that has picked this up is unrelated, which suggests some more tests in this area may be pertinent – although in some ways they would really be testing the behaviour of a third-party component (`ext-dom` or `Masterminds\html5-ptp`), albeit indirectly.

Reply:
> Masterminds/html5-php actually has rules defining non-Boolean attributes to be output with ="" but these do not kick in when the disable_html_ns option is set.

How much did you dig into that? Enough to submit a PR to them? Doesn't seem too obvious to me how `disable_html_ns` affects the code from a quick glance.

Reply:
[`OutputRules::nonBooleanAttribute`](https://github.com/Masterminds/html5-php/blob/2.7.0/src/HTML5/Serializer/OutputRules.php#L371) decides this.

If there is no namespace, it always returns `false`, which means the attribute is considered Boolean and we get the empty attribute syntax.

However, if there is a namespace, as I'm sure you've found, our XPath expressions always fail because, e.g. `xmlns:p` != `p`...